File: native/ws/ws.go
package ws

import (
	_ "embed"
	"fmt"

	"github.com/OutOfBedlam/jsh"
	"github.com/dop251/goja"
	"github.com/dop251/goja_nodejs/eventloop"
	"github.com/gorilla/websocket"
)

//go:embed ws.js
var wsJS string

func Module(rt *goja.Runtime, module *goja.Object) {
	// Export native functions to embedded JS module
	module.Set("NewWebSocket", NewNativeWebSocket)

	// Run the embedded JS module code
	rt.Set("module", module)
	_, err := rt.RunString(fmt.Sprintf(`(()=>{%s})()`, wsJS))
	if err != nil {
		panic(err)
	}
}

func NewNativeWebSocket(obj *goja.Object) *WebSocket {
	return &WebSocket{
		obj: obj,
	}
}

type WebSocket struct {
	obj *goja.Object
}

func (ws *WebSocket) Connect(addr string) (*websocket.Conn, error) {
	if s, _, err := websocket.DefaultDialer.Dial(addr, nil); err != nil {
		return nil, err
	} else {
		return s, nil
	}
}

func (ws *WebSocket) Send(conn *websocket.Conn, typ int, message any) error {
	switch val := message.(type) {
	case string:
		return conn.WriteMessage(typ, []byte(val))
	case []byte:
		return conn.WriteMessage(typ, val)
	default:
		return fmt.Errorf("unsupported message type: %T", val)
	}
}

func (ws *WebSocket) emit(loop *eventloop.EventLoop, eventType string, args ...any) {
	jsh.Emit(loop, ws.obj, eventType, args...)
}

func (ws *WebSocket) Start(conn *websocket.Conn, loop *eventloop.EventLoop) {
	go ws.run(conn, loop)
}

func (ws *WebSocket) run(conn *websocket.Conn, loop *eventloop.EventLoop) {
	for {
		typ, message, err := conn.ReadMessage()
		if err != nil {
			if !websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
				ws.emit(loop, "close", err)
			} else {
				ws.emit(loop, "error", err)
			}
			return
		}
		data := map[string]any{}
		if typ == websocket.TextMessage {
			data["data"] = string(message)
		} else {
			data["data"] = message
		}
		data["type"] = typ
		ws.emit(loop, "message", data)
	}
}

File: native/ws/ws.js
class WebSocket extends EventEmitter {
    constructor(url) {
        super();
        if (url === undefined || typeof url !== 'string') {
            throw new TypeError('URL must be a string, got ' + typeof url);
        }
        this.supportedEvents.push("open", "close", "message", "error");
        this.raw = module.NewWebSocket(this);
        this.readyState =  WebSocket.CONNECTING;
        this.url = url;

        const {eventLoop} = require('process');
        eventLoop.runOnLoop(() => {
            try {
                const conn = this.raw.connect(this.url);
                if (conn instanceof Error) {
                    this.emit('error', conn);
                } else {
                    this.conn = conn;
                    this.readyState = WebSocket.OPEN;
                    this.emit('open');
                    this.raw.start(this.conn, eventLoop);
                }
            }catch(e){
                this.emit('error', new Error(e.value));
            }
        });      
    }

    send(data) {
        if (!this.conn) {
            let err = new Error("websocket is not open");
            this.emit('error', err);
            return;
        }
        try {
            this.raw.send(this.conn, WebSocket.TextMessage, data);
        } catch (err) {
            this.emit('error', err);
        }
    }
    close() {
        this.readyState = this.CLOSING;
        if(this.conn) {
            this.conn.close();
        }
        this.readyState = this.CLOSED;
        this.emit('close');
    }

    static CONNECTING = 0;
    static OPEN = 1;
    static CLOSING = 2;
    static CLOSED = 3;

    static TextMessage = 1;
    static BinaryMessage = 2;
}

module.exports = {
    WebSocket: WebSocket,
}